Add bluetooth host keys unit test

ahmedmoheb-nordic opened the pull request:
This is part of /host/key.c UT projects.
Please review in order to catch review comments as early as possible

ahmedmoheb-nordic commented:
@jori-nordic @kruithofa please review

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_overwrite_oldest.c:
```diff
@@ -0,0 +1,241 @@
+ *   - CONFIG_BT_KEYS_OVERWRITE_OLDEST is enabled
+ *
+ *  Expected behaviour:
+ *   - Valid reference pointer is returned which matches the oldest
```
- NULL pointer is returned.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_overwrite_oldest.c:
```diff
@@ -0,0 +1,241 @@
+	return false;
+}
+
+/* Tests need to be kept in the same order */
```
I really don't like this comment. Refer to the one in `keys_get_addr/src/main.c`.

ahmedmoheb-nordic replied:
This test suite verifies the order at which the oldest key is overwritten.
So, it's actually required to run test cases in the correct order so that we can expect the output.

ahmedmoheb-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -0,0 +1,171 @@
+ *   - Address value matches the one passed to bt_keys_get_addr()
+ *   - The keys pool list is full
+ */
+void test_fill_keys_pool(void)
```
I'm working on that, but I'll create a separate PR for modifications in Keys.c and Keys.h

ahmedmoheb-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -0,0 +1,171 @@
+		ztest_unit_test(test_round_initialization),
+		ztest_unit_test(test_list_is_empty_at_startup),
+		LISTIFY(CONFIG_BT_MAX_PAIRED, REGISTER_SETUP_TEARDOWN, (,), test_fill_keys_pool),
+		ztest_unit_test(test_no_equal_references));
```
I think you got your impression that these are just a setup functions because the same functions and names are used in other functions UT projects like `bt_keys_foreach_bond` and `bt_keys_foreach_type`.

I can agree with your proposal and call these functions directly without considering them test cases, except for this one in `bt_keys_get_addr/src/main.c`.

This one is the only real one that is testing `test_get_non_existing_keys` and may be the name `test_fill_keys_pool` is a bit misleading and gives a wrong impression.

So, I agree to apply your suggestions in other projects because they are actually used as a setup functions.
Here, we can just re-name `test_fill_keys_pool` to `test_get_non_existing_keys` because this is its main job and filling the keys pool can be considered as a side-effect of it.

jori-nordic replied:
so I think I got half of my point over (which is nice). But I still think you shouldn't be relying on the side effects of previous test cases: each individual UT should be runnable by itself in isolation and you should be able to shuffle them without encountering issues.

ahmedmoheb-nordic replied:
RFC, I think we need to settle on the definition of `UT` in your comment.

In my opinion, It should be used to refer to a `test suite` which is composed of one or more test functions.
Running the `test suite` verifies a certain behavior and it may be required to run test functions in a certain order.

I think `test suites` should be independent from each other and their order shouldn't affect the result.
But a test function inside a `test suite` can be dependent on the output from the previous one.

jori-nordic replied:
I guess that can work, if you abuse the `suite` to be the UT and the `tests` inside it (except one) to be setup-teardown functions.

In that case, could you rename each `test_` fn that is actually setting up the state for the real logic test `setup_`? same for `teardown_` fns. That way, the reader (which could be us in a few years) will understand what's happening where, and know not to reorder the functions not prefixed by `test_`.

More consensus on order-dependent tests being a bad practice:
https://stackoverflow.com/a/53383608
https://softwareengineering.stackexchange.com/a/368803
top of that comment https://softwareengineering.stackexchange.com/a/335433

ahmedmoheb-nordic replied:
> In that case, could you rename each `test_` fn that is actually setting up the state for the real logic test `setup_`?

Unfortunately, that isn't possible, all test functions that will run using the framework must starts wit `test_`.
This can be found here https://docs.zephyrproject.org/3.1.0/develop/test/ztest.html

You can only setup a single setup function to be called before running the test function.

Anyway, I extracted the setup functions outside the `test suite` and they are now explicitly called before running the `test suite` as you previously suggested

ahmedmoheb-nordic committed:
tests: bluetooth: host: Add mocks for keys.c

Add required mocks to be able to compile and test /bluetooth/host/keys.c

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

ahmedmoheb-nordic committed:
tests: bluetooth: host: Add UT for bt_keys_get_addr()

Unit test project for bt_keys_get_addr().
This part of subsys/bluetooth/host/buf.c unit testing

Signed-off-by: Ahmed Moheb <ahmed.moheb@nordicsemi.no>

jori-nordic commented on tests/bluetooth/host/host_mocks/print_utils.c:
```diff
@@ -0,0 +1,89 @@
+	return str;
+}
+
+const char *bt_hex_real(const void *buf, size_t len)
```
why can't we use `bluetooth/common/log.c`? Also, where are you using this, is it for visualization or error messages in the UTs?

ahmedmoheb-nordic replied:
Ok, `bluetooth/host/keys.c` uses `BT_DBG(fmt, ...)` which brings `log` module definitions directly.

Firstly, I tried to add  `bluetooth/common/log.c`, but there were a lot of other dependencies and compilations issues that I preferred to avoid as I'm not really interested in 'log.c' itself.

Regarding the UTs, these functions are just used to print a readable strings when some asserts fail.

jori-nordic commented on tests/bluetooth/host/keys/mocks/hci_core_expects.h:
```diff
@@ -0,0 +1,23 @@
+ *  Validate expected behaviour when bt_unpair() isn't called
+ *
+ *  Expected behaviour:
+ *   - bt_unpair() not to called at all
```
to not be called

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -0,0 +1,173 @@
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_pair_lut) == ARRAY_SIZE(returned_keys_refs));
+
+/*
+ *  Test list is empty
```
what list, could you be more specific?

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -0,0 +1,173 @@
+/* Setup test variables */
+static void unit_test_setup(void)
+{
+	zassert_true((params_it < (ARRAY_SIZE(testing_id_addr_pair_lut))),
```
this is wrong, you should rather increment `params_it` before making the bounds check

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -0,0 +1,173 @@
+	 */
+	ztest_test_suite(bt_keys_get_addr_get_non_existing_keys,
+			 ztest_unit_test(test_list_is_empty_at_startup),
+			 LISTIFY(CONFIG_BT_MAX_PAIRED, REGISTER_SETUP_TEARDOWN, (,),
```
After discussion with @alwa-nordic , we thought it'd be better to explicitly do the iteration over the test vectors in the test cases themselves. It shouldn't be that hard, just a for-loop pretty much.
We don't _need_ one test case result for each vector x testcase. That way the looping is also more explicit for the reader.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -0,0 +1,173 @@
+ *   - ID value matches the one passed to bt_keys_get_addr()
+ *   - Address value matches the one passed to bt_keys_get_addr()
+ */
+void test_get_non_existing_keys(void)
```
I would rather name it `test_populate_key_pool`.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -0,0 +1,173 @@
+	bt_addr_le_t *addr = current_params_vector->addr;
+
+	returned_key = bt_keys_get_addr(id, addr);
+	/* 'params_it' is incremented before running the test case */
```
this should not be needed anymore when you do the vector iteration in the test function itself.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -0,0 +1,173 @@
+{
+	reset_global_test_round_params();
+
+	/* This testsuite should setup the conditions for subsequent tests
```
In https://github.com/zephyrproject-rtos/zephyr/pull/48676#discussion_r947825731 you agree that test suites should be independent, but here you state the opposite. Which is true?

ahmedmoheb-nordic replied:
fixed!, test suites should be independent from each other

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_invalid_values.c:
```diff
@@ -0,0 +1,135 @@
+ *   - CONFIG_BT_KEYS_OVERWRITE_OLDEST is enabled
+ *
+ *  Expected behaviour:
+ *   - An assertion is triggered and execution stops
```
an assertion fails

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_invalid_values.c:
```diff
@@ -0,0 +1,135 @@
+ */
+void test_null_address_reference(void)
+{
+	expect_assert();
```
why do we have an `expect_assert()` here and not in the previous fns, that should also be raising an assert (per their docstring).

ahmedmoheb-nordic replied:
Yes, it's a bit misleading, it even took me sometime to figure out how come the test is passing if I'm not using it.
I found out that I added it in the callback functions assigned to `bt_conn_foreach_fake.custom_fake`.

Anyway, I moved it just before calling `bt_keys_get_addr()` to make it more readable.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_no_overwrite.c:
```diff
@@ -0,0 +1,46 @@
+	zassert_true(returned_key == NULL, "bt_keys_get_addr() returned a non-NULL reference");
+}
+
+static bool pragma_overwrite_not_enabled(const void *s)
```
I'd name that one `overwrite_oldest_disabled`. Pragma is misleading because it means something in the context of the preprocessor. If you want something, you could use `predicate`, that's what the function signature defines.

ahmedmoheb-nordic replied:
Your point is totally valid, I just followed the name used in the test framework getting started guide, but it's misleading.
I changed it to `predicate` as used in the current version of the guide.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_no_overwrite.c:
```diff
@@ -0,0 +1,46 @@
+	return true;
+}
+
+ztest_register_test_suite(bt_keys_get_addr_full_list_no_overwrite, pragma_overwrite_not_enabled,
```
it looks like this is deprecated (it is defined in `ztest_test_deprecated.h`). 
Could you update your ztest api calls so they don't use deprecated macros/functions?
For this one, I think the new one is `ZTEST_SUITE`.

ahmedmoheb-nordic replied:
Yes, was afraid to catching this one, it's going to take some effort :).
I already started using the new APIs with next modules

ahmedmoheb-nordic replied:
The point is that using the new APIs wasn't available before Torsten's changes, and I've started working on the `Keys` module while the first PR was in review.

So, I'm not sure if I should apply these review comments to the `buf` tests as well or not, please advise?

jori-nordic replied:
yes could you do that? the PR should be straightforward to review as it's just an API change.

ahmedmoheb-nordic replied:
> yes could you do that? the PR should be straightforward to review as it's just an API change.

Unfortunately, it doesn't work in that way.

In order to use new APIs, you will need to add `CONFIG_ZTEST_NEW_API=y` to `prj.conf` and once you do that, all deprecated definitions won't be available and you need to switch all APIs to the new ones.

ahmedmoheb-nordic replied:
My point is that it's doable, but it takes time because the new ones and the deprecated ones aren't identical in terms of suites and tests setup functions and signatures.

I'm currently testing how much effort is required to switch to the new ones and I will starting working on the `buf` module after applying the review comments here to the other commits.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_no_overwrite.c:
```diff
@@ -0,0 +1,46 @@
+ *  Test adding extra (ID, Address) pair while the keys pool list is full
+ *
+ *  Constraints:
+ *   - Keys pool list is full
```
I don't see where this is done, I would expect a setup function run before or at the start of the test suite that actually does this.
If you're relying on state from a previous state, that's not ok.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_overwrite_oldest.c:
```diff
@@ -0,0 +1,302 @@
+/* Pointer to the current set of oldest testing parameter */
+struct id_addr_pair const *oldest_params_vector;
+
+/* This call back will be called when bt_unpair() is called with the same parameters
```
with the same parameters as what

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_overwrite_oldest.c:
```diff
@@ -0,0 +1,302 @@
+struct id_addr_pair const *oldest_params_vector;
+
+/* This call back will be called when bt_unpair() is called with the same parameters
+ * It will clear the key slot with the assigned to the 'id'
```
This sentence doesn't make sense to me. Reading the callback, what it does is get the first `(id, addr)` match and clear it. So it doesn't just depend on `id`, but also `addr`.
I would phrase it `Clears the first key slot that matches its arguments`.

ahmedmoheb-nordic replied:
I followed your suggestion to remove the comment

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_overwrite_oldest.c:
```diff
@@ -0,0 +1,302 @@
+/* This call back will be called when bt_conn_foreach() is called with the same parameters
+ * It will act as if there are some active connections
+ */
+static void bt_conn_foreach_custom_fake_key_slot_0_in_use(int type, bt_conn_foreach_cb func,
```
Please either pick the end or the start for your `custom_fake` pre/suf-fix, but not the middle like this.
I would be in favor of removing it entirely, or just having a `fake_` prefix.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_overwrite_oldest.c:
```diff
@@ -0,0 +1,302 @@
+}
+
+/*
+ *  Test adding extra (ID, Address) pair while the keys pool list is full
```
`Add a new (ID, Address) pair while the keys pool is full and all the keys are in use.`
Same for the next testcase with `and all the keys aren't in use`.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_overwrite_oldest.c:
```diff
@@ -0,0 +1,302 @@
+ *
+ *  Expected behaviour:
+ *   - Valid reference pointer is returned which matches the oldest
+ *     reference at index 1
```
index 0

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_overwrite_oldest.c:
```diff
@@ -0,0 +1,302 @@
+ *  Constraints:
+ *   - Keys pool list is full
+ *   - Key at slot 0 is used with a connection
+ *   - Oldest key isn't used with a connection
```
`oldest key (slot 2)` isn't used with a connection`
I don't really like this test case because it implicitly depends on the state set by the previous one (if I understood correctly).
Could you group the two into a single one?

jori-nordic reviewed (changes requested):
I see some of my comments could make it a pain to update all the other commits, and it will be hard to see if they're resolved across all the files (2000 lines changed).

Could we focus on the `bt_keys_get_addr` suites only in this PR (ie, could you remove the other commits), get it in good shape and merge it, and then review the rest? I think it'll be easier and faster.

ahmedmoheb-nordic commented:
> I see some of my comments could make it a pain to update all the other commits, and it will be hard to see if they're resolved across all the files (2000 lines changed).
> 
> Could we focus on the `bt_keys_get_addr` suites only in this PR (ie, could you remove the other commits), get it in good shape and merge it, and then review the rest? I think it'll be easier and faster.

Ok, let's do that.

ahmedmoheb-nordic commented:
> I see some of my comments could make it a pain to update all the other commits, and it will be hard to see if they're resolved across all the files (2000 lines changed).
> 
> Could we focus on the `bt_keys_get_addr` suites only in this PR (ie, could you remove the other commits), get it in good shape and merge it, and then review the rest? I think it'll be easier and faster.

As we agreed, I kept only the first 2 commits related to the `mocks` and `bt_keys_get_addr` test suites.
Also, I added a set of `fixup` commits on top of that according to your review commits for easier changes tracking.

Please check and let me know.

jori-nordic commented on tests/bluetooth/host/keys/mocks/keys_help_utils.c:
```diff
@@ -17,6 +17,24 @@ void clear_key_pool(void)
+	struct id_addr_pair const *params_vector;
+
+	zassert_true(check_key_pool_is_empty(),
+		     "List isn't empty, make sure to run this test just after a fresh start");
```
This function should only be in the test suite setup.

ahmedmoheb-nordic replied:
Ok, I wasn't fully satisfied with it as well.

My point was that this function will usually be used while setting up the test and it's behavior won't be checked.
So, I wasn't sure if this function should implicitly clear the keys pool to ensure correct behavior or it shouldn't alter the current state of the keys pool.

jori-nordic commented on tests/bluetooth/host/keys/mocks/keys_help_utils.c:
```diff
@@ -17,6 +17,24 @@ void clear_key_pool(void)
+		id = params_vector->id;
+		addr = params_vector->addr;
+		refs[it] = bt_keys_get_addr(id, addr);
+		zassert_true(refs[it] != NULL, "bt_keys_get_addr() returned a non-valid reference");
```
I would suggest `Failed to add key to the keys pool`.
The existing message is technically correct but won't help much when debugging (and the debugger has forgotten the weird behavior of `bt_keys_get_addr`).

ahmedmoheb-nordic replied:
From your previous comment here, I thought that you aren't in favor of using any `zassert_` checks outside the test suite setup functions or test cases.

jori-nordic replied:
I'm reviewing commit by commit so I haven't made it to the end yet if you have changed this further. 
I think the assert is fine, it's just the assert message I was being nitpicky about.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -8,6 +8,9 @@
+#include "host_mocks/assert.h"
+#include "mocks/conn.h"
+#include "mocks/hci_core.h"
 #include "host_mocks/print_utils.h"
```
nitpick: can you group the includes by folder?

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -100,6 +114,16 @@ void test_no_equal_references(void)
 }
 
+/* Setup test variables */
+static void every_test_setup(void *f)
```
`test_case_setup`? What you have is also fine.

jori-nordic replied:
looks like this macro will allow us to get rid of the `LISTIFY` hack in the buf.c UTs.

ahmedmoheb-nordic replied:
Ok, I was just trying to distinguish test cases setup functions from test suite setup ones.

The names used by the test framework `ztest_suite_setup_t` `ztest_suite_before_t` `ztest_suite_after_t` `ztest_suite_teardown_t` are not very descriptive to me.

jori-nordic replied:
It's actually better defined in `ztest.rst`.
- `setup`: run once at the beginning of the suite
- `before`: run at the beginning of each test case
- `after`: run at the end of each test case
- `teardown`: run once at the end of the suite

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -134,36 +158,31 @@ void test_get_key_by_matched_id_and_address(void)
+void test_main(void)
+{
+	/* Only startup suite will run. */
+	all_startup_checks_executed = false;
```
IMO, 
```c
zassert_true(check_key_pool_is_empty(),
		     "List isn't empty, make sure to run this test just after a fresh start");			     "List isn't empty, make sure to run this test just after a fresh start");
```
Should be part of the test case setup (`every_test_setup`), right after `clear_key_pool`. or even _in_ `clear_key_pool`.

ahmedmoheb-nordic replied:
To make it simple, let's agree on not using `zassert_` checks in sources outside the project source files.
That leaves us with one choice to make that check part of the test cases setup.

jori-nordic replied:
yep I was going to suggest that as an alternative.

jori-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/test_suite_full_list_invalid_values.c:
```diff
@@ -41,6 +47,16 @@ static void bt_conn_foreach_custom_fake_data_ref_null(int type, bt_conn_foreach_
 }
 
+/* Setup test variables */
+static void every_test_setup(void *f)
```
this seems identical to the one in main.c.
I think you can remove the `static` and declare it as an `extern`. If you did this because the `ZTEST_SUITE` macro has a limitation, then it's ok.

ahmedmoheb-nordic replied:
I'm not sure if the test framework has such limitation, but I have a concern here.

- It's just a coincidence here that those functions match, but I prefer to keep the setup function local to the same source file that has the test case definition as part of making each test suite and its test cases independent from the others

jori-nordic replied:
That's a fine argument (y)

jori-nordic commented:
I think the PR looks much better now, names are more descriptive, has been migrated to the new ztest API, etc. I only have some minor nits, and then we can hopefully merge.

ahmedmoheb-nordic commented:
> I think the PR looks much better now, comments are more descriptive, has been migrated to the new ztest API, etc. I only have some minor nits, and then we can hopefully merge.

I added some more fixup commits to address your latest comments.

jori-nordic commented on tests/bluetooth/host/keys/mocks/keys_help_utils.c:
```diff
@@ -23,15 +23,20 @@ void fill_key_pool_by_id_addr(const struct id_addr_pair src[], int size, struct
 
-	zassert_true(check_key_pool_is_empty(),
-		     "List isn't empty, make sure to run this test just after a fresh start");
+	if (!check_key_pool_is_empty()) {
```
I'm thinking instead of having something with user interaction (printing to terminal), you could just return a negative error code (of your choosing in errno.h) and assert in the caller that it returns 0.

ahmedmoheb-nordic replied:
> I'm thinking instead of having something with user interaction (printing to terminal), you could just return a negative error code (of your choosing in errno.h) and assert in the caller that it returns 0. edit: I mean, you can leave the printout, thats cool for debugging, but also add a return value.

Ok, it's a simple change

tejlmand reviewed (approved):
lgtm
